src/UCTNodePointer.cpp

```diff
@@ -45,52 +46,71 @@ UCTNodePointer::UCTNodePointer(std::int16_t vertex, float policy) {
     std::memcpy(&i_policy, &policy, sizeof(i_policy));
 
     m_data =  (static_cast<std::uint64_t>(i_policy)  << 32)
-            | (static_cast<std::uint64_t>(i_vertex) << 16) | 1ULL;
+            | (static_cast<std::uint64_t>(i_vertex) << 16);
 }
 
 UCTNodePointer& UCTNodePointer::operator=(UCTNodePointer&& n) {
-    if (is_inflated()) {
-        delete read_ptr();
-    }
-    m_data = n.m_data;
-    n.m_data = 1;
+    auto nv = std::atomic_exchange(&n.m_data, INVALID);
+    auto v = std::atomic_exchange(&m_data, nv);
 
+    if (is_inflated(v)) {
+        delete read_ptr(v);
+    }
     return *this;
 }
 
 void UCTNodePointer::inflate() const {
-    if (is_inflated()) return;
-    m_data = reinterpret_cast<std::uint64_t>(
-        new UCTNode(read_vertex(), read_policy()));
+    while (true) {
+        auto v = m_data.load();
+        if (is_inflated(v)) return;
+
+        auto v2 = reinterpret_cast<std::uint64_t>(
+            new UCTNode(read_vertex(v), read_policy(v))
+        ) | POINTER;
+        bool success = m_data.compare_exchange_strong(v, v2);
+        if (success) {
+            return;
+        } else {
+            // this means that somebody else also modified this instance.
```

Comment:
Could they have done anything else beside inflate it?

Reply:
UCTNodePointer should be considered read only if it is not inflated.  That was the assumption - this is supposed to be mimicking  lazy initialization.